osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -0,0 +1,376 @@
+                distanceToPrev * (float)Math.Sin(currentObjectInfo.AngleRad)
+            );
+
+            posRelativeToPrev = getRotatedVector(prevObjectInfo.EndPositionRandomised, posRelativeToPrev);
+
+            currentObjectInfo.AngleRad = (float)Math.Atan2(posRelativeToPrev.Y, posRelativeToPrev.X);
+
+            var position = Vector2.Add(prevObjectInfo.EndPositionRandomised, posRelativeToPrev);
+
+            // Move hit objects back into the playfield if they are outside of it,
+            // which would sometimes happen during big jumps otherwise.
+            position.X = MathHelper.Clamp(position.X, 0, OsuPlayfield.BASE_SIZE.X);
+            position.Y = MathHelper.Clamp(position.Y, 0, OsuPlayfield.BASE_SIZE.Y);
+
+            currentObjectInfo.PositionRandomised = position;
+        }
+
+        /// <summary>
+        /// Moves the <see cref="Slider"/> and all necessary nested <see cref="OsuHitObject"/>s into the <see cref="OsuPlayfield"/> if they aren't already.
+        /// </summary>
+        private void moveSliderIntoPlayfield(ref Slider slider, ref RandomObjectInfo currentObjectInfo)
+        {
+            var oldPos = new Vector2(slider.Position.X, slider.Position.Y);
+
+            // Min. distances from the slider's position to the playfield border
+            var minMargin = new MarginPadding();
+
+            foreach (var hitObject in slider.NestedHitObjects.Where(o => o is SliderTick || o is SliderEndCircle))
+            {
+                if (!(hitObject is OsuHitObject osuHitObject))
+                    continue;
+
+                var relativePos = Vector2.Subtract(osuHitObject.Position, slider.Position);
+
+                minMargin.Left = Math.Max(minMargin.Left, -relativePos.X);
+                minMargin.Right = Math.Max(minMargin.Right, relativePos.X);
+                minMargin.Top = Math.Max(minMargin.Top, -relativePos.Y);
+                minMargin.Bottom = Math.Max(minMargin.Bottom, relativePos.Y);
+            }
+
+            if (slider.Position.X < minMargin.Left)
+                slider.Position = new Vector2(minMargin.Left, slider.Position.Y);
+            else if (slider.Position.X + minMargin.Right > OsuPlayfield.BASE_SIZE.X)
+                slider.Position = new Vector2(OsuPlayfield.BASE_SIZE.X - minMargin.Right, slider.Position.Y);
+
+            if (slider.Position.Y < minMargin.Top)
+                slider.Position = new Vector2(slider.Position.X, minMargin.Top);
+            else if (slider.Position.Y + minMargin.Bottom > OsuPlayfield.BASE_SIZE.Y)
+                slider.Position = new Vector2(slider.Position.X, OsuPlayfield.BASE_SIZE.Y - minMargin.Bottom);
+
+            currentObjectInfo.PositionRandomised = slider.Position;
+            currentObjectInfo.EndPositionRandomised = slider.EndPosition;
+
+            var shift = Vector2.Subtract(slider.Position, oldPos);
+
+            shiftNestedObjects(slider, shift);
```

Comment:
This should probably be moved to the same method as the other call to the same function. It's hard to follow if it's done in two separate places.

Reply:
I've moved this for you, but it needs an inline comment explaining why these two shifts are required in the first place.